File: Java/1.Two-Sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer,Integer> map = new HashMap();
        int[] ret = new int[2];
        for(int i = 0; i<nums.length; ++i) {
            int number = target- nums[i];
            if(map.containsKey(number)) {
                return new int[]{map.get(number), i};
            }
            map.put(nums[i], i);
        }
        return new int[]{};
    }
}


/*
HashMap比较占用内存，不过以时间换空间，时间复杂度降低很多
*/

File: C++/1.Two-Sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        int vSize = nums.size();
        map<int,int> m;
        for(int i = 0; i < vSize;++i){
            auto iter = m.find(target-nums[i]);
            if(iter != m.end()){
                return {iter->second, i};
            }
            // 找不到再insert，否则先insert [3,3] 6这种测试例就会失败
            m.insert({nums[i],i});  
        }
        return {};
    }
};

/*tip:
1. 暴力解决, 两个for循环, 时间复杂度达到了O(n^2), 不符合要求;
2. 用空间换时间, 把对应数及索引创建一个map;
3. 从目标数里减去一个数，查找另外一个数是否在map中, over;

*/
